### PR TITLE
Dynamic font size 2.0

### PR DESCRIPTION
**Pull request by peteward**
implements #209 

Revised implementation for dynamic font sizing:
- grow and shrink font to fit vertical space
- improve UX for managing dynamic font sizing

### UI changes

* font size now becomes disabled when you enable dynamic font sizing - it doesn't make sense anymore as the adjustment is dynamic.
* instead the calculated dynamic size is updated in real-time into the disabled box
* this also improves the performance of dynamic font calculations because in most cases you will be typing content and the change will be relative to what has already been typed. 
* disabling dynamic font sizing retains the last calculated dynamic size
* limits on minimum and maximum font sizes have been removed to improve UX, which can be confusing if you want to do so by typing, but changes don't take effect when deleting values.
* instead we won't do any dynamic calculations if the values are invalid because min > max, we use the last known font size value instead.

https://github.com/pdfme/pdfme/assets/7068515/4614533d-9552-4fd9-b0aa-31b727fd1c28

### Generator and common code
* consolidated the `getSplittedLines` function into common so that it can be used by both the UI and the generator.
* fixed a bug in this logic (#218) which could leave lines over the width limit and cause accidental wrapping on rendering 
* consolidated some calculation methods and clarified variables names where different units were being used (pt vs mm)

**Tests to be updated if happy with the approach**

### CODE REVIEW

File: packages/common/src/font.ts [hunk 1]
```diff
@@ -104,23 +103,18 @@ export const heightOfFontAtSize = (fontKitFont: FontKitFont, fontSize: number) =
...
-  const numberOfCharacters = textContent.length;
-  return (numberOfCharacters - 1) * textCharacterSpacing;
+const calculateCharacterSpacing = (textContent: string, textCharacterSpacing: number) => {
+  return (textContent.length - 1) * textCharacterSpacing;
 };
 
-const calculateTextWidthInMm = (textContent: string, textWidth: number, textCharacterSpacing: number) =>
-  (textWidth + calculateCharacterSpacing(textContent, textCharacterSpacing)) * DEFAULT_PT_TO_MM_RATIO;
-
+export const widthOfTextAtSize = (text: string, fontKitFont: FontKitFont, fontSize: number, characterSpacing: number) => {
```
Comment (peteward) [inline packages/common/src/font.ts:110]: we now include the character spacing as part of text width calculations

File: packages/common/src/font.ts [hunk 2]
```diff
@@ -136,58 +130,156 @@ export const getFontKitFont = async (textSchema: TextSchema, font: Font) => {
...
-    textContentRows.forEach((line) => {
-      const textWidth = widthOfTextAtSize(line, fontKitFont, size);
-      const textWidthInMm = calculateTextWidthInMm(line, textWidth, characterSpacingCount);
-
-      totalWidthInMm = Math.max(totalWidthInMm, textWidthInMm);
-      totalHeightInMm += textHeightInMm;
+    const textHeightInMm = pt2mm(textHeight * lineHeight);
+
+    textContentRows.forEach((paragraph) => {
+      const lines = getSplittedLines(paragraph, {
```
Comment (peteward) [inline packages/common/src/font.ts:239]: in order to get the full height of the text we need to both split it on line breaks and split by the box width

File: packages/generator/src/helper.ts [hunk 3]
```diff
@@ -130,14 +134,7 @@ export const getEmbeddedPagesAndEmbedPdfBoxes = async (arg: {
...
 
-const mm2pt = (mm: number): number => {
-  // https://www.ddc.co.jp/words/archives/20090701114500.html
-  const ptRatio = 2.8346;
-
-  return parseFloat(String(mm)) * ptRatio;
-};
-
-const getSchemaSizeAndRotate = (schema: Schema) => {
+const convertSchemaDimensionsToPt = (schema: Schema) => {
```
Comment (peteward) [inline packages/generator/src/helper.ts:137]: renamed this function to clarify that it's doing a unit conversion (it took me a lot of time to notice this!)

File: packages/ui/src/components/Schemas/TextSchema.tsx [hunk 4]
```diff
@@ -17,37 +17,39 @@ import { FontContext } from '../../contexts';
...
   const [fontAlignmentValue, setFontAlignmentValue] = useState<number>(0);
 
+  useEffect(() => {
+    getFontKitFont(schema, font).then((fontKitFont) => {
+      const fav = getFontAlignmentValue(fontKitFont, schema.fontSize ?? DEFAULT_FONT_SIZE);
+      setFontAlignmentValue(fav);
+    });
+  }, [schema, schema.width, font]);
 
   useEffect(() => {
```
Comment (peteward) [inline packages/ui/src/components/Schemas/TextSchema.tsx:33]: the `useEffect` was split into 2 because listening on a superset of attributes led the dynamic font sizing code into an infinite loop (or at least an extra execution). 
This is because we're adjusting the `schema.fontSize` value so we need to avoid listening for this change or we end up looping.

File: packages/common/src/constants.ts [hunk 5]
```diff
@@ -4,10 +4,10 @@ export const DEFAULT_ALIGNMENT = 'left';
...
 export const DEFAULT_CHARACTER_SPACING = 0;
 export const DEFAULT_FONT_COLOR = '#000';
-export const DEFAULT_TOLERANCE = 1;
-export const DEFAULT_FONT_SIZE_ADJUSTMENT = 0.25;
-export const DEFAULT_PT_TO_PX_RATIO = 1.333;
-export const DEFAULT_PT_TO_MM_RATIO = 0.3528;
+export const FONT_SIZE_ADJUSTMENT = 0.25;
+export const PT_TO_PX_RATIO = 1.333;
+export const PT_TO_MM_RATIO = 0.3528;
+export const MM_TO_PT_RATIO = 2.8346; // https://www.ddc.co.jp/words/archives/20090701114500.html
```
[the remark this answers is not in the document]
Comment (hand-dot) [reply packages/common/src/constants.ts:10]: 👍

**Comment by hand-dot** [comment]
@peteward I'll check within holiday.

File: packages/generator/src/helper.ts (hunk 3 above)
Comment (hand-dot) [reply packages/generator/src/helper.ts:137]: Sorry...

File: packages/common/src/font.ts [hunk 6]
```diff
@@ -136,58 +130,156 @@ export const getFontKitFont = async (textSchema: TextSchema, font: Font) => {
...
-  const fontSize = _fontSize || DEFAULT_FONT_SIZE;
+  // For very long lines with no whitespace use the original overPos
+  return overPosTmp > 0 ? overPosTmp : overPos - 1;
+};
+
+/**
+ * Recursively splits the line at getSplitPosition.
+ * If there is some leftover, split the rest again in the same manner.
+ */
+export const getSplittedLines = (textLine: string, calcValues: FontWidthCalcValues): string[] => {
```
[the remark this answers is not in the document]
Comment (hand-dot) [reply packages/common/src/font.ts:187]: Nice.

[the remark this answers is not in the document]
Comment (hand-dot) [reply packages/common/src/font.ts:187]: @peteward 
Wait, moved it to "common," but isn't it being used in the UI?

This is fine for now, but if it's not going to be used commonly, wouldn't it have been better to move it to a different file?

If we split functions, it might become difficult to delete them in the future, and there's a possibility that they could become unused functions. Additionally, considering the Custom Schema that we will implement in the future, I think it would be better to gather all the processes related to this TextSchema in one place.

**Comment by hand-dot** [comment]
@peteward 
I  have not yet been able to actually run this revised version locally to confirm its functionality, but I have completed the code check. 
Apart from the following point, there are no issues with the code:
https://github.com/pdfme/pdfme/pull/222#discussion_r1289425930
You may choose to fix this comment or leave it as is.

> Tests to be updated if happy with the approach
It's a great fix. Please update the tests.

File: packages/common/src/font.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (peteward) [reply packages/common/src/font.ts:187]: Originally this function was only used in the generator. Now it is used by both the generator and the UI, so I moved it to common/font.
Let me know if you think there is a more appropriate place.

[the remark this answers is not in the document]
Comment (hand-dot) [reply packages/common/src/font.ts:187]: Ah, sorry, it's being used in 'calculateDynamicFontSize', I missed that.

**Comment by peteward** [comment]
Thanks!

I have a busy day but will try to update the tests as soon as possible

**Comment by hand-dot** [comment]
I actually ran it and checked the operation, but there's something I've been thinking about, and I want to exchange opinions.

I think one of the major revision points this time is that we've made the Dynamic font size fit the vertical space, but
when we enforce vertical space, I think there might be instances where the behavior isn't as expected, like when you want to retain line breaks in multi-line text. What do you think?

Specifically, in the case of an address label like the one below, I definitely want to input it in two lines, but the possibility of it becoming three lines might arise due to this revision.
CleanShot 2023-08-11 at 11 33 52@2x


Don't you think an option for Dynamic font size like "fit: vertical or horizontal" is necessary?

**Comment by hand-dot** [comment]
In other words,

With fit: vertical, you can prevent other text from becoming abnormally small when one of the multiple lines becomes abnormally long.
With fit: horizontal, you can prevent the text from breaking into a new line on its own.

I believe that these settings should be used differently on a case-by-case basis.

**Comment by peteward** [comment]
Hi @hand-dot,

Yes, this makes sense.

We are also working with addresses, although I would likely only use dynamic font size to reduce size, not increase it. In which case the vertical fit is fine. However, I concede that this definitely a valid use case.

Is this a current problem for you? Will this PR as it is now be a negative change for you? 

It would be great if we could merge this as it is now (once tests are complete), but if it does cause a problem for you then I will try to add the functionality first.  I don't think it will be too difficult (I hope!)

**Comment by hand-dot** [comment]
@peteward 

We cannot release this change as it is now. This is because if dynamic font sizing fits vertically or horizontally, and if its behavior changes arbitrarily, it will confuse users.

Fortunately, version 2.0 has been released not too long ago, so we want to add new settings to dynamic font sizing at this time. Until now, it has been fitting horizontally, but from now on, we will make it so that users can set it to fit vertically.

I'm sorry for the trouble, but could you implement it by referring to the code and UI below?
```plaintext
export const TextSchema = CommonSchema.extend({
  type: z.literal(SchemaType.Enum.text),
  alignment: Alignment.optional(),
  fontSize: z.number().optional(),
  fontName: z.string().optional(),
  fontColor: z.string().optional(),
  backgroundColor: z.string().optional(),
  characterSpacing: z.number().optional(),
  lineHeight: z.number().optional(),
  dynamicFontSize: z.object({
    max: z.number(),
    min: z.number(),
    fit: z.enum(['vertical', 'horizontal']).optional(),
  }).optional(),
});
```

CleanShot 2023-08-12 at 16 45 45@2x

*If dynamicFontSize.fit is undefined, it should be the usual fit: horizontal.

**Comment by peteward** [comment]
Hi @hand-dot,

I have added a lot of tests, especially around `getSplittedLines` as there were some edge cases that I have now caught.
I have also added an integration test PDF for dynamic font sizing.

If you are happy with this PR, then can we merge it and I will follow up with another one to add the choice of vertical or horizontal fit?

**Comment by hand-dot** [comment]
@peteward 
OK, I've merged it for now.
The tests are failing, so I want you to fix that first.

After that, let's make it so we can choose between fit: vertical and horizontal.
Thank you as always.
